File: services/orchestrator/internal/pool/trim.lua
-- trim.lua — rút MỘT pod thừa khỏi pool:free, nguyên tử cùng phép so trần.
--
-- KEYS[1] = pool:free
-- ARGV[1] = POOL_TARGET (trần)
--
-- Trả tên pod vừa rút; trả nil khi pool KHÔNG vượt trần (ca bình thường, không
-- phải lỗi — Go rẽ nhánh "hết việc" trên redis.Nil, cùng khuôn ErrPoolEmpty của
-- claim.lua).
--
-- ⛔ VÌ SAO PHẢI LÀ LUA CHỨ KHÔNG PHẢI `LLEN` RỒI `RPOP` TỪ GO.
-- Giữa hai lệnh đó một claim có thể lấy đi một pod (claim.lua LMOVE từ đầu
-- TRÁI). Pop tiếp khi đó là rút xuống DƯỚI trần, rồi vòng replenish sau tạo lại
-- — churn do chính ta gây ra, và log sẽ nói "đã rút pod thừa" trong lúc pool
-- đang THIẾU. Ở đây phép so và phép pop nằm trong cùng một lượt nguyên tử nên
-- trạng thái mà ta quyết định trên đó không thể đổi giữa hai bước.
--
-- KHÁC claim.lua ở một điểm quyết định hình dạng của file này: script này ghi
-- ĐÚNG MỘT lệnh. Bài học 1.A-2 ("Redis Lua có isolation, KHÔNG có rollback") chỉ
-- sinh ra nghĩa vụ tự-hoàn-tác cho script NHIỀU ghi — ở đây không có bước nào ở
-- giữa để hỏng, nên không có gì phải hoàn tác. Đừng thêm pcall/undo vào đây cho
-- "đối xứng với claim.lua": nó sẽ là mã chết giả dạng phòng thủ.
--
-- RPOP (đầu PHẢI) chứ không LPOP: claim.lua LMOVE từ đầu TRÁI, nên rút ở đầu
-- phải (a) giữ nguyên FIFO của D6 — pod cũ nhất vẫn ra trước nên pod hỏng vẫn lộ
-- sớm — và (b) bỏ đúng pod MỚI NHẤT, tức pod vừa bị một manager thứ hai tạo
-- thừa. Rút ở đầu trái là tranh trực tiếp với đường claim của người dùng và bỏ
-- đi đúng pod đã ấm lâu nhất.
--
-- RPOP CŨNG CHÍNH LÀ PHÉP GIÀNH QUYỀN SỞ HỮU, cùng luật với `LREM` của reaper
-- tầng 4: trả về một tên = ta vừa rút tên đó khỏi pool ⇒ không claim nào còn
-- grab được nó ⇒ ta được phép xoá pod. Trả nil = không có gì của ta để xoá.

local target = tonumber(ARGV[1])
if target == nil or target < 0 then
  -- Trần vô nghĩa thì KHÔNG rút gì. Fail-closed ở đây quan trọng hơn bình
  -- thường: nhánh này xoá pod, nên một ARGV hỏng mà rơi vào "trần = 0" sẽ vét
  -- sạch warm-pool mỗi vòng và mọi log đều nói "đã rút pod thừa".
  return redis.error_reply('trim: POOL_TARGET khong hop le')
end

if redis.call('LLEN', KEYS[1]) <= target then
  return nil
end

return redis.call('RPOP', KEYS[1])

File: services/orchestrator/internal/pool/trim.go
package pool

import (
	"context"
	_ "embed"
	"errors"
	"fmt"
	"log/slog"
	"strconv"

	"github.com/redis/go-redis/v9"

	"github.com/Nghaiz/DevOps_Learning_Platform/services/orchestrator/internal/k8s"
	"github.com/Nghaiz/DevOps_Learning_Platform/services/shared/rediskeys"
)

//go:embed trim.lua
var trimLua string

// trimScript — cùng nghi thức EVALSHA-trước-EVAL-sau với claimScript (bài học
// A5: Redis restart làm trống script cache và trả NOSCRIPT).
var trimScript = redis.NewScript(trimLua)

// maxTrimPerCycle chặn số pod rút trong MỘT vòng.
//
// Ở trạng thái bình thường phần thừa là 1 (hai manager cùng thấy pool hụt và
// mỗi bên dựng một pod). Trần này là lưới cho ca `pool:free` phình bất thường —
// và chính ca đó là triệu chứng ta muốn NHÌN THẤY qua `dlp_pool_free_size` chứ
// không phải thứ nên biến một vòng replenish thành một tràng lời gọi apiserver.
// Cùng lý lẽ với freeBatch/quarantineBatch/claimedBatch của reaper.
const maxTrimPerCycle = 10

// trimSurplus rút phần pod ấm VƯỢT trần POOL_TARGET rồi xoá chúng.
//
// ⛔ VÌ SAO NHÁNH NÀY TỒN TẠI — `POOL_TARGET` TỪNG CHỈ LÀ SÀN, KHÔNG PHẢI TRẦN.
// `replenishOnce` chỉ bơm thêm khi thiếu; không nhánh nào rút bớt khi thừa. Đo
// được 2026-08-11 sau `helm upgrade`: `dlp_pool_free_size = 2` với
// `POOL_TARGET=1`, và con số 2 Ở LẠI VĨNH VIỄN ⇒ trần session đồng thời tụt
// 3 → 2 (D16: `trần = quota_hiệu_lực − |pool:free|`) mà không lỗi nào nói vì
// sao. Nguồn của phần thừa là `maxSurge` của rollout (hai manager cùng sống
// ~13s), NHƯNG `values.yaml` đặt orchestrator `replicaCount: 2` nên trên mọi
// deploy không dùng `values-selfhost.yaml` thì HAI manager sống THƯỜNG TRỰC —
// tức đây không phải hiện tượng thoáng qua lúc rollout.
//
// Chốt đường (a) tự-sửa thay vì (b) leader-election: hai manager vẫn có thể
// cùng tạo, nhưng phần thừa bị rút ở vòng sau nên trần luôn quay về đúng. Giá
// là CHURN (một pod bị tạo rồi xoá mỗi lần hai manager đua) — không phải rò.
func (m *Manager) trimSurplus(ctx context.Context) error {
	var errs []error

	for i := 0; i < maxTrimPerCycle; i++ {
		if ctx.Err() != nil {
			return errors.Join(append(errs, ctx.Err())...)
		}

		name, err := trimScript.Run(ctx, m.rdb,
			[]string{rediskeys.PoolFree},
			strconv.Itoa(m.target),
		).Text()
		switch {
		case errors.Is(err, redis.Nil):
			// Không còn pod thừa. Đây là ca BÌNH THƯỜNG và là đường ra duy nhất
			// của vòng lặp ở trạng thái khoẻ.
			return errors.Join(errs...)
		case err != nil:
			return errors.Join(append(errs, fmt.Errorf("pool: chạy trim.lua: %w", err))...)
		}

		// Tới đây ta SỞ HỮU `name`: RPOP đã rút nó khỏi pool:free nên không claim
		// nào còn grab được nó.
		m.met.PoolTrimmedTotal.Inc()
		m.log.Warn("rút pod ấm THỪA khỏi pool — nhiều warm-pool manager đang cùng bơm",
			slog.String("pod", name),
			slog.Int("pool_target", m.target))

		if err := m.deleteSurplus(ctx, name); err != nil {
			errs = append(errs, err)
		}
	}

	return errors.Join(errs...)
}

// deleteSurplus xoá một pod thừa mà ta đã giành được quyền sở hữu.
//
// ⛔ `DEL pod:{name}` TRƯỚC, `pods.Delete` SAU — THỨ TỰ NÀY LÀ LUẬT.
//
// Đảo lại thì một lượt `Delete` hỏng (apiserver chớp, 503 lúc rollout
// control-plane) để lại một pod CÓ hash `pod:{name}`, KHÔNG nằm trong list nào,
// và KHÔNG session nào trỏ tới — mà không tầng reaper nào phủ được ca đó: tầng
// 2a đòi hash VẮNG, 2b đòi có `session:{id}`, 2c quét `pool:claimed`, tầng 3
// quét `pool:quarantine`, tầng 4 quét `pool:free`. Đó là điểm mù thứ SÁU, và nó
// tránh được bằng thứ tự chứ không cần thêm một tầng nữa.
//
// Xoá hash trước thì lượt hỏng để lại một pod HASH-LESS trong `dlp-sandbox` với
// label `app=sandbox` — tức đúng định nghĩa mồ côi của tầng 2a, sẽ được dọn sau
// `orphanGrace`. Chọn chế độ hỏng nằm TRONG vùng phủ của một tầng đã có.
func (m *Manager) deleteSurplus(ctx context.Context, name string) error {
	podKey, err := rediskeys.Pod(name)
	if err != nil {
		return err
	}

	if err := m.rdb.Del(ctx, podKey).Err(); err != nil && !errors.Is(err, redis.Nil) {
		// Không xoá pod khi hash còn: xem thứ tự ở trên. Vòng sau thử lại —
		// nhưng tên đã ra khỏi pool:free nên không ai claim phải nó, và tầng 4
		// không thấy nó nữa. Pod trở thành việc của tầng 2a sau orphanGrace.
		return fmt.Errorf("pool: DEL %s (pod thừa %q): %w", podKey, name, err)
	}

	if err := m.pods.Delete(ctx, name, 0); err != nil && !k8s.IsNotFound(err) {
		return fmt.Errorf("pool: xoá pod thừa %q: %w", name, err)
	}
	return nil
}
